### PR TITLE
[BIOX-1946] Copy Paste Support

### PR DESCRIPTION
**Pull request by VDykyi**
Implemented copy/paste on vpaint side

### CODE REVIEW

File: src/VAC/Scene.cpp [hunk 1]
```diff
@@ -113,6 +113,16 @@ void Scene::setCanvasDefaultValues()
...
 }
 
+bool Scene::isSelectionNotEmpty()
+{
+    Layer * layer = activeLayer();
+    if(layer)
+    {
+        return layer->vac()->selectedCells().count() > 0;
+    }
+    return false;
```
Comment (marcus-sonestedt) [inline src/VAC/Scene.cpp]: return layer && layer -> ....;  ? :)

File: src/VAC/VectorAnimationComplex/VAC.cpp [hunk 2]
```diff
@@ -348,15 +350,47 @@ QMap<int,int> VAC::import(VAC * other, bool selectImportedCells)
...
 
     // Take ownership of all cells
-    for(Cell * c: qAsConst(ordering))
-    {
-        int oldID = c->id();
-        copyOfOther->removeCell_(c);
-        insertCellLast_(c);
+    KeyVertex* firstKeyVertex = nullptr;
+    double leftX = global()->scene()->width();
+    double topY = global()->scene()->height();
```
Comment (marcus-sonestedt) [inline src/VAC/VectorAnimationComplex/VAC.cpp]: auto

File: src/VAC/VectorAnimationComplex/VAC.cpp [hunk 3]
```diff
@@ -348,15 +350,47 @@ QMap<int,int> VAC::import(VAC * other, bool selectImportedCells)
...
+        copyOfOther->removeCell_(cell);
+        insertCellLast_(cell);
         if(selectImportedCells)
-            addToSelection(c,false);
-        res[oldID] = c->id();
+            addToSelection(cell, true);
+        res[oldID] = cell->id();
+
+        KeyVertex* currentKeyVertex = cell->toKeyVertex();
+        if (currentKeyVertex != nullptr)
```
Comment (marcus-sonestedt) [inline src/VAC/VectorAnimationComplex/VAC.cpp]: can be combined

File: src/VAC/VectorAnimationComplex/VAC.cpp [hunk 4]
```diff
@@ -348,15 +350,47 @@ QMap<int,int> VAC::import(VAC * other, bool selectImportedCells)
...
+        {
+            double currentX = currentKeyVertex->pos()[0];
+            double currentY = currentKeyVertex->pos()[1];
```
Comment (marcus-sonestedt) [inline src/VAC/VectorAnimationComplex/VAC.cpp]: auto again. :)

File: src/VAC/VectorAnimationComplex/VAC.h [hunk 5]
```diff
@@ -477,6 +477,9 @@ public slots:
     // Transform tool
     TransformTool transformTool_;
     friend class TransformTool;
+
+    double pasteDeltaX_;
+    double pasteDeltaY_;
```
Comment (marcus-sonestedt) [inline src/VAC/VectorAnimationComplex/VAC.h:482]: I feel safer if these are zero-initialized when declared. :)

File: src/VAC/View.cpp [hunk 6]
```diff
@@ -240,24 +246,24 @@ int View::decideClicAction()
...
-            return SPLIT_ACTION;
-        }
+//        if( global()->toolMode() == Global::SELECT &&
+//             mouse_LeftButton_ &&
+//             !mouse_AltWasDown_ &&
+//             mouse_ControlWasDown_ &&
+//             !mouse_ShiftWasDown_)
+//        {
+//            return SPLIT_ACTION;
+//        }
```
Comment (marcus-sonestedt) [inline src/VAC/View.cpp]: add comment why this is commented out, or if it's not needed in future, remove it

File: src/VAC/View.cpp [hunk 7]
```diff
@@ -240,24 +246,24 @@ int View::decideClicAction()
...
         // Paint
-        if( global()->toolMode() == Global::PAINT &&
-            mouse_LeftButton_ &&
-            !mouse_AltWasDown_ &&
-            !mouse_ControlWasDown_ &&
-            !mouse_ShiftWasDown_)
-        {
-            return PAINT_ACTION;
-        }
+//        if( global()->toolMode() == Global::PAINT &&
```
Comment (marcus-sonestedt) [inline src/VAC/View.cpp]: same

**Review by marcus-sonestedt — APPROVED** [review]

File: src/VAC/Scene.cpp (hunk 1 above)
Comment (VDykyi) [reply src/VAC/Scene.cpp]: Removed

File: src/VAC/VectorAnimationComplex/VAC.cpp (hunk 2 above)
Comment (VDykyi) [reply src/VAC/VectorAnimationComplex/VAC.cpp]: Done

File: src/VAC/VectorAnimationComplex/VAC.cpp (hunk 3 above)
Comment (VDykyi) [reply src/VAC/VectorAnimationComplex/VAC.cpp]: done

File: src/VAC/VectorAnimationComplex/VAC.cpp (hunk 4 above)
Comment (VDykyi) [reply src/VAC/VectorAnimationComplex/VAC.cpp]: done

File: src/VAC/VectorAnimationComplex/VAC.h (hunk 5 above)
Comment (VDykyi) [reply src/VAC/VectorAnimationComplex/VAC.h:482]: I added initializations in constructors

File: src/VAC/View.cpp (hunk 6 above)
Comment (VDykyi) [reply src/VAC/View.cpp]: removed

File: src/VAC/View.cpp (hunk 7 above)
Comment (VDykyi) [reply src/VAC/View.cpp]: removed

**Pull request merged**